### PR TITLE
Wazo 2638: Migrate xivo-lib-python from Marshmallow 3.0.0b14 to 3.10.0

### PR DESCRIPTION
**Pull request by nissou91**

### CODE REVIEW

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
## Critical
* You cannot leave tests commented in the tox file
* Since marshmallow is not compatible anymore with python2
  * You need to skip marshmallow tests when running pytest with python2
  * You need to raise a error when importing these files with python2

## Warning
* Learn how to work with git efficiently
  * rebase your commits
  * Make commit message useful, don't duplicate it
  * no need to have JIRA ID before each commit, branch name already have it
  * Don't include useless file in the git tree, rebase and fixup !
 
* Follow wazo workflow about merge master into your branch vs rebase
 
## Questions
* Why adding `unknown=EXCLUDE` everywhere? It was a workaround for `Nested` field to fix WAZO-1729, but all schemas should inherit from base schema which already define this property
* Did you read marshmallow changelog? I don't see any of mentioned deprecated/warning changes
  * `field.fail` is deprecated and should use `make_error`instead
  * No need to pass `many` or `partial` to decorator, they are always returned (we should remove them and leave default behavior)
  * `field.IP` is now implemented and we should base our implementation on it 


## Nice to have
Can you write a small summary of the impacting changes about wazo usage of this library, for other devs? You can add it to the JIRA ticket description

File: integration_tests/suite/test_service_discovery.py [hunk 1]
```diff
@@ -255,4 +255,4 @@ def _get_message(self):
         try:
             return self.messages.get(timeout=30)
         except queue.Empty:
-            self.fail('Should have received a message')
+            self.make_error('Should have received a message')
```
Comment (fblackburn1) [inline integration_tests/suite/test_service_discovery.py:258]: Why did you change this method?

File: requirements.txt [hunk 2]
```diff
@@ -7,6 +7,7 @@ netifaces==0.10.4
 psycopg2==2.7.7
 pyyaml==3.13
 python-consul==1.1.0
-marshmallow==3.0.0b14
+marshmallow==3.0.0b14; python_version <= '2.7'
+marshmallow==3.10.0; python_version >= '3.6'
```
Comment (fblackburn1) [inline requirements.txt:11]: Nice trick :+1:

File: xivo/mallow/fields.py [hunk 3]
```diff
@@ -1,23 +1,33 @@
...
+        UUID as _UUID,
+    )
+except Exception:
+    import sys
+
+    python_major_version = sys.version_info.major
+    if python_major_version == 2:
+        raise ImportError(
+            "Marshamallow library is incompatible with Python version %s" % sys.version
+        )
```
Comment (fblackburn1) [inline xivo/mallow/fields.py:29]: What if import doesn't trigger `Exception`? Can we use this kind of condition instead? (not tested)
```suggestion
import sys

python_major_version = sys.version_info.major
if python_major_version == 2:
    raise ImportError(
        "Marshamallow library is incompatible with Python version %s" % sys.version
    )

from marshmallow.fields import (
    Boolean as _Boolean,
    Constant as _Constant,
    Date as _Date,
    DateTime as _DateTime,
    Dict as _Dict,
    Email as _Email,
    Field as _Field,
    Float as _Float,
    Integer as _Integer,
    List as _List,
    Nested as _Nested,
    String as _String,
    TimeDelta as _TimeDelta,
    URL as _URL,
    UUID as _UUID,
)
```

File: xivo/mallow/tests/test_fields.py [hunk 4]
```diff
@@ -21,13 +33,14 @@ class AllFieldsSchema(Schema):
     dict_ = fields.Dict()
     email = fields.Email()
     list_ = fields.List(fields.String())
-    nested = fields.Nested('TestSchema')
+    nested = fields.Nested(lambda: TestSchema())
```
Comment (fblackburn1) [inline xivo/mallow/tests/test_fields.py:36]: Why using lambda? According the changelog the "String" usage is not deprecated
https://marshmallow.readthedocs.io/en/stable/changelog.html#id21

File: xivo/mallow/tests/test_validate.py [hunk 5]
```diff
@@ -13,7 +14,17 @@
...
+try:
+    from marshmallow import Schema, ValidationError
+except Exception:
+
+    python_major_version = sys.version_info.major
+    if python_major_version == 2:
+        raise ImportError(
+            "Marshamallow library is incompatible with Python version %s" % sys.version
+        )
+
```
Comment (fblackburn1) [inline xivo/mallow/tests/test_validate.py:27]: Same comment than above

File: xivo/mallow/validate.py [hunk 6]
```diff
@@ -1,19 +1,27 @@
...
+        URL as _URL,
+    )
+except Exception:
+    import sys
+
+    python_major_version = sys.version_info.major
+    if python_major_version == 2:
+        raise ImportError(
+            "Marshamallow library is incompatible with Python version %s" % sys.version
+        )
```
Comment (fblackburn1) [inline xivo/mallow/validate.py:24]: same comment than above

File: xivo/mallow_helpers.py [hunk 7]
```diff
@@ -44,6 +53,7 @@ class ListSchema(marshmallow.Schema):
     search = fields.String(missing=None)
 
     class Meta:
+        register = False
```
Comment (fblackburn1) [inline xivo/mallow_helpers.py:56]: Why not using default register? It's very convenient

File: xivo/tests/test_mallow_helpers.py [hunk 8]
```diff
@@ -16,12 +17,22 @@
...
+
+try:
+    from marshmallow import fields, ValidationError
+except Exception:
+
+    python_major_version = sys.version_info.major
+    if python_major_version == 2:
+        raise ImportError(
+            "Marshamallow library is incompatible with Python version %s" % sys.version
+        )
```
Comment (fblackburn1) [inline xivo/tests/test_mallow_helpers.py:29]: same comment than above

File: xivo/tests/test_pubsub.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def test_unsubscribe_when_never_subscribed(self):
         try:
             self.pubsub.unsubscribe(SOME_TOPIC, callback)
         except Exception:
-            self.fail('unsubscribe should not raise exceptions')
+            self.make_error('unsubscribe should not raise exceptions')
```
Comment (fblackburn1) [inline xivo/tests/test_pubsub.py:65]: same comment than above. Why?

File: xivo/tests/test_pubsub.py [hunk 10]
```diff
@@ -77,7 +77,7 @@ def publish_when_nobody_subscribed(self):
         try:
             self.pubsub.publish(SOME_TOPIC, SOME_MESSAGE)
         except Exception:
-            self.fail('publish should not raise exceptions')
+            self.make_error('publish should not raise exceptions')
```
Comment (fblackburn1) [inline xivo/tests/test_pubsub.py:80]: same comment than above, why?

**Review by fblackburn1 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 2 above)
Comment (nissou91) [reply requirements.txt:11]: Thanks :)

File: xivo/tests/test_pubsub.py (hunk 9 above)
Comment (nissou91) [reply xivo/tests/test_pubsub.py:65]: same mistake as mentioned above

File: xivo/tests/test_pubsub.py (hunk 10 above)
Comment (nissou91) [reply xivo/tests/test_pubsub.py:80]: same mistake as mentioned above

File: xivo/mallow/tests/test_validate.py (hunk 5 above)
Comment (nissou91) [reply xivo/mallow/tests/test_validate.py:27]: Your suggestion did not work; I guess the change I made is more convenient.

Comment (nissou91) [reply xivo/mallow/tests/test_validate.py:27]: I got the following output when applying your suggestion:
```
______________ ERROR collecting xivo/mallow/tests/test_fields.py _______________
ImportError while importing test module '/root/src/github.com/wazo-platform/xivo-lib-python/xivo/mallow/tests/test_fields.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
xivo/mallow/tests/test_fields.py:22: in <module>
    from .. import fields
xivo/mallow/fields.py:9: in <module>
    "Marshamallow library is incompatible with Python version %s" % sys.version
E   ImportError: Marshamallow library is incompatible with Python version 2.7.16 (default, Oct 10 2019, 22:02:15) 
E   [GCC 8.3.0]
```

Comment (fblackburn1) [reply xivo/mallow/tests/test_validate.py]: Yeah and this is the expected behavior. Because pytest to *enter* in each module to collect tests. According their documentation, you can use this technique:
https://docs.pytest.org/en/6.2.x/example/pythoncollection.html#customizing-test-collection
But I commit fix

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
Remaining comments:

* `field.IP` is now implemented and we should base our implementation on it
* Can you write a small summary of the impacting changes about wazo usage of this library, for other devs? You can add it to the JIRA ticket description
